libuavcan/include/libuavcan/transport/media/can.hpp

```diff
@@ -0,0 +1,336 @@
+         * payload length supported for the DLC.
+         */
+        constexpr std::uint8_t dlc_to_length_lookup[] = {0, 1, 2, 3, 4, 5, 6, 7, 8, 12, 16, 20, 24, 32, 48, 64};
+
+        static_assert(std::extent<decltype(dlc_to_length_lookup)>::value == 16,
+                      "dlc_to_length_lookup table is malformed.");
+
+        const auto dlc_value = static_cast<std::underlying_type<FrameDLC>::type>(dlc);
+        if (dlc_value < std::extent<decltype(dlc_to_length_lookup)>::value)
+        {
+            return dlc_to_length_lookup[dlc_value];
+        }
+        else
+        {
+            return 0;
+        }
+    }
+
+    std::uint32_t id;  ///< CAN ID with flags (above)
+    std::uint8_t  data[MtuBytesParam];
+
+private:
+    FrameDLC dlc;  ///< Data Length Code
+
+public:
+    Frame()
+        : id(0)
+        , dlc(FrameDLC::CodeForLength0)
+    {
+        std::fill(data, data + MtuBytesParam, std::uint8_t(0));
+    }
+
+    Frame(std::uint32_t can_id, const std::uint8_t* can_data, FrameDLC in_dlc)
+        : id(can_id)
+        , dlc(in_dlc)
+    {
+        const std::uint_fast8_t data_len = dlcToLength(dlc);
+        LIBUAVCAN_ASSERT(nullptr_t != can_data);
+        LIBUAVCAN_ASSERT(data_len <= MtuBytesParam);
+        std::copy(can_data, can_data + static_cast<std::uint8_t>(data_len), this->data);
+    }
+
+    FrameDLC getDLC() const
+    {
+        return dlc;
+    }
+
+    void setDataLength(std::uint_fast8_t data_length)
+    {
+        dlc = lengthToDlc(data_length);
+    }
+
+    std::uint_fast8_t getDataLength() const
+    {
+        return dlcToLength(dlc);
+    }
+
+    bool operator!=(const Frame& rhs) const
+    {
+        return !operator==(rhs);
+    }
+
+    bool operator==(const Frame& rhs) const
+    {
+        return (id == rhs.id) && (dlc == rhs.dlc) && equal(data, data + dlcToLength(dlc), rhs.data);
+    }
+
+    bool operator<(const Frame& other) const
+    {
+        return this->priorityLowerThan(other);
+    }
+
+    bool operator>(const Frame& other) const
+    {
+        return this->priorityHigherThan(other);
+    }
+
+    bool isExtended() const
+    {
+        return id & FlagEFF;
+    }
+
+    bool isRemoteTransmissionRequest() const
+    {
+        return id & FlagRTR;
+    }
+
+    bool isErrorFrame() const
+    {
+        return id & FlagERR;
+    }
+
+    /**
+     * CAN frame arbitration rules, particularly STD vs EXT:
+     *     Marco Di Natale - "Understanding and using the Controller Area Network"
+     *     https://www.amazon.com/dp/B00F5USEOE
+     */
+    bool priorityHigherThan(const Frame& rhs) const
+    {
+        const std::uint32_t clean_id     = id & MaskExtID;
+        const std::uint32_t rhs_clean_id = rhs.id & MaskExtID;
+
+        /*
+         * STD vs EXT - if 11 most significant bits are the same, EXT loses.
+         */
+        const bool ext     = id & FlagEFF;
+        const bool rhs_ext = rhs.id & FlagEFF;
+        if (ext != rhs_ext)
```

Comment:
I recall that my PX4 profiling exercises revealed that the TX queue management code spends a lot of time inside this method. Granted, that was before the new AVL-tree based queue contributed by Theodoros, but even with that tree this will likely remain a hot path. There exist very cheap ways of optimizing it:

- Drop support for standard frames (CAN 2.0A, 11-bit ID) from libuavcan. UAVCAN does not use CAN 2.0A anyway (at least not currently) so why bother handling them? We could just put an additional constraint at the *media layer interface* saying that standard frames shall not pass through. While we're at it, the RTR and error frames should be let go also.
- Use branching hints which are available in most compilers nowadays. We would define something like this for GCC and Clang, allowing the user to override them:

```c++
#define LIBUAVCAN_LIKELY(x)     __builtin_expect((x), true)
#define LIBUAVCAN_UNLIKELY(x)   __builtin_expect((x), false)
```

Our empirical studies revealed that on a Cortex-M4F target equipped with a four-cycle 128-bit ROM prefetcher those can lead to substantial performance improvements (same study also revealed that they can cause an equally dramatic performance degradation when the stated prediction does not match the actual runtime behavior).

Would you consider adding the macros like above into libuavcan and sprinkling them around the code where pertinent?

Reply:
>Drop support for standard frames (CAN 2.0A, 11-bit ID) from libuavcan. UAVCAN does not use CAN 2.0A anyway (at least not currently) so why bother handling them? We could just put an additional constraint at the media layer interface saying that standard frames shall not pass through. While we're at it, the RTR and error frames should be let go also.

Completely agree. I don't think this is anything that we need in the standard though, right? This is just libuavcan saying "we only pass through valid UAVCAN frames. Frames with 11-bit IDs, remote frames, and error frames are not valid UAVCAN frames".

>Use branching hints which are available in most compilers nowadays. We would define something like this for GCC and Clang, allowing the user to override them:

I would like to hold off on this until I can actually obtain profiling data myself.

Reply:
> Completely agree. I don't think this is anything that we need in the standard though, right? This is just libuavcan saying "we only pass through valid UAVCAN frames. Frames with 11-bit IDs, remote frames, and error frames are not valid UAVCAN frames".

Yes.

Reply:
I'm about to post one more change that provides guidance based on this thread.